src/ds/address.h

```diff
@@ -39,4 +42,80 @@ namespace snmalloc
   {
     return reinterpret_cast<T*>(address);
   }
+
+  /**
+   * Test if a pointer is aligned to a given size, which must be a power of
+   * two.
+   */
+  template<size_t alignment>
+  static inline bool is_aligned_block(void* p, size_t size)
+  {
+    assert(bits::next_pow2(alignment) == alignment);
+
+    return ((static_cast<size_t>(address_cast(p)) | size) & (alignment - 1)) ==
+      0;
+  }
+
+  /**
+   * Align a pointer down to a statically specified granularity, which must be a
+   * power of two.
+   */
+  template<size_t granule, typename T = void>
+  inline T* pointer_align_down(void* p)
```

Comment:
`void*` or `T*`,  I guess the thought is that changing alignment will probably change the type?

Reply:
In `largealloc`, `pagemap`, and `pal_open_enclave`, these functions get used on generic `void` or `char` pointers, but in the various slab uses, they are used to find the slab header and so give the type.  It seemed in keeping with `pointer_cast` and more concise than `static_cast<Slab*>(pointer_align_down<SLAB_SIZE>(this))`, for example.